upup/pkg/fi/cloudup/scaleway/cloud.go

```diff
@@ -16,7 +16,393 @@ limitations under the License.
+			}
+		}
+		if server.PrivateIP != nil {
+			cloudInstance.PrivateIP = *server.PrivateIP
+		}
+	}
+
+	return cloudInstanceGroup, nil
+}
+
+func (s *scwCloudImplementation) FindClusterStatus(cluster *kops.Cluster) (*kops.ClusterStatus, error) {
+	klog.V(8).Info("FindClusterStatus is not implemented yet for Scaleway")
+	return nil, nil
+}
+
+func (s *scwCloudImplementation) GetApiIngressStatus(cluster *kops.Cluster) ([]fi.ApiIngressStatus, error) {
+	//TODO(Mia-Cross) implement me
+	panic("implement me")
+}
+
+func (s *scwCloudImplementation) GetClusterServers(clusterName string, serverName *string) ([]*instance.Server, error) {
+	request := &instance.ListServersRequest{
+		Zone: s.zone,
+		Name: serverName,
+		Tags: []string{TagClusterName + "=" + clusterName},
+	}
+	servers, err := s.instanceAPI.ListServers(request, scw.WithAllPages())
+	if err != nil {
+		if serverName != nil {
+			return nil, fmt.Errorf("failed to list cluster servers named %q: %w", *serverName, err)
+		}
+		return nil, fmt.Errorf("failed to list cluster servers: %w", err)
+	}
+	return servers.Servers, nil
+}
+
+func (s *scwCloudImplementation) GetClusterVolumes(clusterName string) ([]*instance.Volume, error) {
+	volumes, err := s.instanceAPI.ListVolumes(&instance.ListVolumesRequest{
+		Zone: s.zone,
+		Tags: []string{TagClusterName + "=" + clusterName},
+	}, scw.WithAllPages())
+	if err != nil {
+		return nil, fmt.Errorf("failed to list cluster volumes: %w", err)
+	}
+	return volumes.Volumes, nil
+}
+
+func (s *scwCloudImplementation) DeleteServer(server *instance.Server) error {
+	srv, err := s.instanceAPI.GetServer(&instance.GetServerRequest{
+		Zone:     s.zone,
+		ServerID: server.ID,
+	})
+	if err != nil {
+		if is404Error(err) {
+			klog.V(4).Infof("delete server %s: instance was already deleted", server.ID)
+			return nil
+		}
+		return err
+	}
+
+	// If the server is running, we turn it off and wait before deleting it
+	if srv.Server.State == instance.ServerStateRunning {
+		_, err := s.instanceAPI.ServerAction(&instance.ServerActionRequest{
+			Zone:     s.zone,
+			ServerID: server.ID,
+			Action:   instance.ServerActionPoweroff,
+		})
+		if err != nil {
+			return fmt.Errorf("delete server %s: error powering off instance: %w", server.ID, err)
+		}
+	}
+	_, err = s.instanceAPI.WaitForServer(&instance.WaitForServerRequest{
+		ServerID: server.ID,
+		Zone:     s.zone,
+	})
+	if err != nil {
+		return fmt.Errorf("delete server %s: error waiting for instance after power-off: %w", server.ID, err)
+	}
+
+	// We delete the server and wait before deleting its volumes
+	err = s.instanceAPI.DeleteServer(&instance.DeleteServerRequest{
+		ServerID: server.ID,
+		Zone:     s.zone,
+	})
+	if err != nil {
+		return fmt.Errorf("delete server %s: error deleting instance: %w", server.ID, err)
+	}
+	for {
+		_, err := s.instanceAPI.GetServer(&instance.GetServerRequest{
+			Zone:     s.zone,
+			ServerID: server.ID,
+		})
+		if is404Error(err) {
+			break
+		}
+	}
```

Comment:
Or we just do a WaitServer which is what this function is for.

Reply:
Does WaitServer works when deleting a server ?

Reply:
Not sure we care that much to wait for the server to be deleted. I guess that, if there's no error on `DeleteServer()`, it will eventually disappear. Also, is there any constraint to stop the server before deleting it?

Reply:
We later delete the volumes attached to the server, there might be a constraint that disallow this action while the server is on. I don't know if the constraint is also present if the server still exists. @Mia-Cross